src/Bicep.Local.Extension/Host/TypeDefinitionBuilder/TypeDefinitionBuilder.cs

```
@@ -0,0 +1,186 @@
+            flags: ResourceFlags.None,
+            functions: null));
+
+    protected virtual TypeBase GenerateForRecord(TypeFactory factory, ConcurrentDictionary<Type, TypeBase> typeCache, Type type)
+    {
+        var typeProperties = new Dictionary<string, ObjectTypeProperty>();
+
+
+
+        foreach (var property in type.GetProperties(BindingFlags.Public | BindingFlags.Instance))
+        {
+            if (visited.Contains(property.PropertyType))
+            {
+                // infinite recursion prevention
+                continue;
+            }
+
+            var annotation = property.GetCustomAttributes<TypeAnnotationAttribute>(true).FirstOrDefault();
+            var propertyType = property.PropertyType;
+
+            TypeBase? typeReference = null;
+
+            if (TryResolveTypeReference(propertyType, annotation, out typeReference)) { }
+            else if (propertyType != typeof(string) && typeof(IEnumerable).IsAssignableFrom(propertyType))// propertyType.IsAssignableTo(typeof(IEnumerable)))
+            {
+                // protect against infinite recursion
+                visited.Add(property.PropertyType);
+
+                Type? elementType = null;
+                if (propertyType.IsArray)
+                {
+                    elementType = propertyType.GetElementType();
+                }
+                else if (propertyType.IsGenericType &&
+                         propertyType.GetGenericTypeDefinition() == typeof(IEnumerable<>))
+                {
+                    elementType = propertyType.GetGenericArguments()[0];
+                }
+
+                if (elementType is null)
+                {
+                    throw new NotImplementedException($"Unsupported collection type {elementType}");
+                }
+
+                if (TryResolveTypeReference(elementType, annotation, out typeReference)) { }
```

Should this throw, rather than be silently ignored?

---

Probably, as long as the type attribute is required we should throw. If it's optional then we may want to fail silently